File: tools/dump_model_stages.py
"""
Dump YOLO26 stage tensors matching C model_forward_ex names (fused Conv+BN like weights export).
Pair with C dump using weights/yolo26.bin exported from the same checkpoint.

Shared input (recommended for parity vs C): one preprocess path (OpenCV), same float tensor for both sides:
  python tools/dump_model_stages.py --write-shared-input runs/shared_input.bin --image tests/data/zidane.jpg
  ./tests/dump_model_stages runs/shared_input.bin runs/zidane_c_stages.bin weights/yolo26.bin
  python tools/dump_model_stages.py --input-bin runs/shared_input.bin --out runs/zidane_py_stages.bin
"""

from __future__ import annotations

import argparse
import importlib.util
import os
import struct
import sys

import cv2
import numpy as np
import torch
import torch.nn as nn

_REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
_DEFAULT_PT = os.path.join(_REPO, "yolo26n.pt")


def _resolve_path(path: str) -> str:
    """If path is relative and missing cwd, try repo root (for yolo26n.pt / weights/...)."""
    if os.path.isfile(path):
        return path
    alt = os.path.join(_REPO, path)
    if os.path.isfile(alt):
        return alt
    return path


def _load_tool_module(mod_name: str, rel_path: str):
    path = os.path.join(_REPO, rel_path)
    spec = importlib.util.spec_from_file_location(mod_name, path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"cannot load {path}")
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


_gl = _load_tool_module("generate_layer_tests", "tools/generate_layer_tests.py")
save_tensor = _gl.save_tensor
_converter = _load_tool_module("converter", "tools/converter.py")
fuse_conv_bn_state_dict = _converter.fuse_conv_bn_state_dict


def read_first_named_tensor(path: str) -> torch.Tensor:
    """Read one record (same layout as save_tensor / C load_named_tensor)."""
    with open(path, "rb") as f:
        (nl,) = struct.unpack("i", f.read(4))
        f.read(nl)  # name
        (nd,) = struct.unpack("i", f.read(4))
        dims = [struct.unpack("i", f.read(4))[0] for _ in range(nd)]
        while len(dims) < 4:
            dims.append(1)
        n, c, h, w = dims[0], dims[1], dims[2], dims[3]
        count = n * c * h * w
        raw = f.read(4 * count)
        if len(raw) != 4 * count:
            raise ValueError("truncated tensor payload")
    arr = np.frombuffer(raw, dtype=np.float32).copy().reshape((n, c, h, w))
    return torch.from_numpy(arr)


def preprocess_rgb_chw01(bgr_uint8: np.ndarray, w: int, h: int) -> torch.Tensor:
    """Linear resize, BGR->RGB, [1,3,H,W] float32 /255 — canonical shared preprocess (OpenCV)."""
    if bgr_uint8.ndim != 3 or bgr_uint8.shape[2] != 3:
        raise ValueError(f"expected HxWx3 BGR, got {bgr_uint8.shape}")
    resized = cv2.resize(bgr_uint8, (w, h), interpolation=cv2.INTER_LINEAR)
    rgb = cv2.cvtColor(resized, cv2.COLOR_BGR2RGB).astype(np.float32) / 255.0
    chw = np.transpose(rgb, (2, 0, 1))
    return torch.from_numpy(chw).unsqueeze(0).contiguous()


def ensure_fused_conv_biases(net: nn.Module, fused_sd: dict[str, torch.Tensor]) -> int:
    """
    Ultralytics Conv uses Conv2d(..., bias=False) + separate BatchNorm. After fuse_conv_bn_state_dict, fused biases
    exist in fused_sd as *.conv.bias but load_state_dict cannot attach them without a bias Parameter. Rebuild those
    Conv2d layers with bias=True and copy weights + fused bias (matches C inference after fold_all_bn).
    """
    upgraded = 0
    for k, bias_tensor in fused_sd.items():
        if not k.endswith(".conv.bias"):
            continue
        path = k[: -len(".bias")]  # e.g. model.0.conv
        try:
            conv = net.get_submodule(path)
        except Exception:
            continue
        if not isinstance(conv, nn.Conv2d) or conv.bias is not None:
            continue
        new_c = nn.Conv2d(
            conv.in_channels,
            conv.out_channels,
            conv.kernel_size,
            stride=conv.stride,
            padding=conv.padding,
            dilation=conv.dilation,
            groups=conv.groups,
            bias=True,
            padding_mode=conv.padding_mode,
        )
        new_c.weight.data.copy_(conv.weight.data)
        new_c.bias.data.copy_(bias_tensor.data.to(device=new_c.weight.device, dtype=new_c.weight.dtype))
        parent_path, _, name = path.rpartition(".")
        parent = net.get_submodule(parent_path) if parent_path else net
        setattr(parent, name, new_c)
        upgraded += 1
    return upgraded


def replace_bn2d_with_identity(module: nn.Module) -> int:
    """Fused weights already include BN; remove running BatchNorm from the forward path (else activations diverge)."""
    n = 0
    for name, child in list(module.named_children()):
        if isinstance(child, nn.BatchNorm2d):
            setattr(module, name, nn.Identity())
            n += 1
        else:
            n += replace_bn2d_with_identity(child)
    return n


def apply_fused_state_dict(net: torch.nn.Module, verbose: bool = False) -> None:
    """Match tools/converter.py + C runtime: fuse_conv_bn, load weights, add conv biases, strip BN from graph."""
    if hasattr(net, "float"):
        net.float()
    net.eval()
    sd = {k: v for k, v in net.state_dict().items() if isinstance(v, torch.Tensor)}
    before = len(sd)
    fused = fuse_conv_bn_state_dict(sd)
    if verbose:
        print(f"fuse_conv_bn_state_dict: {before} -> {len(fused)} tensors")
    missing, unexpected = net.load_state_dict(fused, strict=False)
    if verbose and (missing or unexpected):
        print("load_state_dict strict=False:", "missing", len(missing), "unexpected", len(unexpected))
    nb = ensure_fused_conv_biases(net, fused)
    if verbose:
        print(f"ensure_fused_conv_biases: upgraded {nb} Conv2d layers to bias=True")
    nid = replace_bn2d_with_identity(net)
    if verbose:
        print(f"replace_bn2d_with_identity: replaced {nid} BatchNorm2d with Identity")


def _tensor_out(x):
    if isinstance(x, torch.Tensor):
        return x
    if isinstance(x, (list, tuple)) and x:
        return x[0] if isinstance(x[0], torch.Tensor) else _tensor_out(x[0])
    return None


def predict_once_dump(net: torch.nn.Module, x: torch.Tensor, f) -> None:
    """Mirror ultralytics BaseModel._predict_once; write stage tensors after each layer (names match C)."""
    y: list = []
    save_tensor(f, "stage_00_input", x.detach().cpu())
    for m in net.model:
        if m.f != -1:
            x = y[m.f] if isinstance(m.f, int) else [x if j == -1 else y[j] for j in m.f]
        x = m(x)
        y.append(x if m.i in net.save else None)
        out_t = _tensor_out(x)
        if out_t is not None:
            save_tensor(f, f"stage_{m.i + 1:02d}_buf{m.i}", out_t.detach().cpu())
    final = _tensor_out(x)
    if final is not None:
        save_tensor(f, "stage_25_detect", final.detach().cpu())


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument(
        "--model",
        default=_DEFAULT_PT,
        help="Ultralytics yolo26n.pt (repo root by default; must match weights/yolo26.bin export)",
    )
    p.add_argument("--image", default="tests/data/zidane.jpg")
    p.add_argument("--out", default="runs/zidane_py_stages.bin")
    p.add_argument(
        "--input-bin",
        default=None,
        metavar="PATH",
        help="Load stage_00_input from this .bin (single save_tensor record); use with --write-shared-input output",
    )
    p.add_argument(
        "--write-shared-input",
        default=None,
        metavar="PATH",
        help="Only write shared preprocess (OpenCV) to this .bin as stage_00_input, then exit",
    )
    p.add_argument("-v", "--verbose", action="store_true")
    args = p.parse_args()

    image_path = _resolve_path(args.image)

    if args.write_shared_input:
        if not os.path.isfile(image_path):
            print(f"Image not found: {args.image} (tried {image_path})", file=sys.stderr)
            sys.exit(1)
        bgr = cv2.imread(image_path, cv2.IMREAD_COLOR)
        if bgr is None:
            print(f"cv2.imread failed: {image_path}", file=sys.stderr)
            sys.exit(1)
        x = preprocess_rgb_chw01(bgr, 640, 640)
        out_shared = args.write_shared_input
        if not os.path.isabs(out_shared):
            out_shared = os.path.join(_REPO, out_shared)
        os.makedirs(os.path.dirname(out_shared) or ".", exist_ok=True)
        with open(out_shared, "wb") as f:
            save_tensor(f, "stage_00_input", x.detach().cpu())
        print(f"Wrote shared input {out_shared} (use as first arg to tests/dump_model_stages, and --input-bin here)")
        return

    model_path = _resolve_path(args.model)
    out_path = args.out if os.path.isabs(args.out) else os.path.join(_REPO, args.out)

    if not os.path.isfile(model_path):
        print(f"Model not found: {args.model} (tried {model_path})", file=sys.stderr)
        sys.exit(1)

    if args.input_bin:
        input_bin = _resolve_path(args.input_bin)
        if not os.path.isfile(input_bin):
            print(f"Input bin not found: {args.input_bin} (tried {input_bin})", file=sys.stderr)
            sys.exit(1)
        x = read_first_named_tensor(input_bin)
        if tuple(x.shape) != (1, 3, 640, 640):
            print(f"Expected input shape (1,3,640,640), got {tuple(x.shape)}", file=sys.stderr)
            sys.exit(1)
    else:
        if not os.path.isfile(image_path):
            print(f"Image not found: {args.image} (tried {image_path})", file=sys.stderr)
            sys.exit(1)
        bgr = cv2.imread(image_path, cv2.IMREAD_COLOR)
        if bgr is None:
            print(f"cv2.imread failed: {image_path}", file=sys.stderr)
            sys.exit(1)
        x = preprocess_rgb_chw01(bgr, 640, 640)

    from ultralytics import YOLO

    ym = YOLO(model_path)
    inner = ym.model
    inner.eval()
    apply_fused_state_dict(inner, verbose=args.verbose)

    os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)
    with open(out_path, "wb") as f:
        with torch.no_grad():
            predict_once_dump(inner, x, f)

    print(f"Wrote {out_path}")


if __name__ == "__main__":
    main()

File: tests/dump_model_stages.c
/*
 * Dump full-model stage tensors to a single .bin (same format as tools/generate_layer_tests.py).
 * Default weights: weights/yolo26.bin (export from yolo26n.pt via tools/converter.py — same checkpoint as Python dump).
 *
 * argv[1] may be an image (jpg/png/...) or a single-tensor .bin from:
 *   python tools/dump_model_stages.py --write-shared-input runs/shared_input.bin --image tests/data/zidane.jpg
 * so C and Python use identical stage_00_input (OpenCV preprocess on the Python side).
 *
 * Build (from repo root, after `make` produces CORE objects):
 *   $(CC) $(CFLAGS) -DNO_LOGGING -Iinclude -Ithird_party tests/dump_model_stages.c \
 *     build/tensor.o build/utils.o build/layers.o build/detection.o build/detect.o build/model.o -o tests/dump_model_stages -lm
 */

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "model.h"
#include "tensor.h"
#include "utils.h"

static int path_ends_with_bin(const char* path) {
    size_t n = strlen(path);
    return n >= 4 && strcmp(path + n - 4, ".bin") == 0;
}

static status_t load_input_from_tensor_bin(const char* path, tensor_t* out_input) {
    FILE* fp = fopen(path, "rb");
    if (!fp) return ERROR_FILE_NOT_FOUND;
    char name[256];
    status_t st = load_named_tensor(fp, name, out_input);
    fclose(fp);
    if (st != SUCCESS) return st;
    if (out_input->dims[0] != 1 || out_input->dims[1] != 3 || out_input->dims[2] != 640 ||
        out_input->dims[3] != 640) {
        tensor_free(out_input);
        return ERROR_INVALID_FORMAT;
    }
    return SUCCESS;
}

static void preprocess(tensor_t* input_tensor, const uint8_t* rgb_buffer, int w, int h) {
    float* data = input_tensor->data;
    int c_stride = h * w;
    for (int y = 0; y < h; y++) {
        for (int x = 0; x < w; x++) {
            const uint8_t* pixel = rgb_buffer + (y * w + x) * 3;
            data[0 * c_stride + y * w + x] = pixel[0] / 255.0f;
            data[1 * c_stride + y * w + x] = pixel[1] / 255.0f;
            data[2 * c_stride + y * w + x] = pixel[2] / 255.0f;
        }
    }
}

static uint8_t* load_rgb_resized(const char* path, int out_w, int out_h) {
    int iw = 0, ih = 0, n = 0;
    unsigned char* img = stbi_load(path, &iw, &ih, &n, 3);
    if (!img) {
        fprintf(stderr, "stbi_load failed: %s\n", path);
        return NULL;
    }
    uint8_t* out = malloc((size_t)out_w * out_h * 3);
    if (!out) {
        stbi_image_free(img);
        return NULL;
    }
    for (int y = 0; y < out_h; y++) {
        for (int x = 0; x < out_w; x++) {
            float fx = (x + 0.5f) * (float)iw / (float)out_w - 0.5f;
            float fy = (y + 0.5f) * (float)ih / (float)out_h - 0.5f;
            int sx = (int)fx;
            int sy = (int)fy;
            if (sx < 0) sx = 0;
            if (sy < 0) sy = 0;
            if (sx >= iw) sx = iw - 1;
            if (sy >= ih) sy = ih - 1;
            int sx1 = sx + 1 < iw ? sx + 1 : sx;
            int sy1 = sy + 1 < ih ? sy + 1 : sy;
            float tx = fx - (float)sx;
            float ty = fy - (float)sy;
            for (int c = 0; c < 3; c++) {
                float v00 = img[(sy * iw + sx) * 3 + c];
                float v10 = img[(sy * iw + sx1) * 3 + c];
                float v01 = img[(sy1 * iw + sx) * 3 + c];
                float v11 = img[(sy1 * iw + sx1) * 3 + c];
                float v = v00 * (1 - tx) * (1 - ty) + v10 * tx * (1 - ty) + v01 * (1 - tx) * ty + v11 * tx * ty;
                out[(y * out_w + x) * 3 + c] = (uint8_t)(v < 0 ? 0 : (v > 255 ? 255 : v));
            }
        }
    }
    stbi_image_free(img);
    return out;
}

int main(int argc, char** argv) {
    const int W = 640, H = 640;
    const char* weights = "weights/yolo26.bin";
    const char* image_path = "tests/data/zidane.jpg";
    const char* out_path = "runs/zidane_c_stages.bin";

    if (argc >= 2) image_path = argv[1];
    if (argc >= 3) out_path = argv[2];
    if (argc >= 4) weights = argv[3];

    model_t model;
    tensor_t input_tensor;
    memset(&input_tensor, 0, sizeof(input_tensor));

    int use_bin_input = path_ends_with_bin(image_path);
    uint8_t* rgb = NULL;

    if (use_bin_input) {
        if (model_create(&model, W, H) != SUCCESS) {
            fprintf(stderr, "model_create failed\n");
            return 1;
        }
        if (load_input_from_tensor_bin(image_path, &input_tensor) != SUCCESS) {
            fprintf(stderr, "load_input_from_tensor_bin failed: %s\n", image_path);
            model_destroy(&model);
            return 1;
        }
    } else {
        rgb = load_rgb_resized(image_path, W, H);
        if (!rgb) return 1;
        if (model_create(&model, W, H) != SUCCESS) {
            free(rgb);
            fprintf(stderr, "model_create failed\n");
            return 1;
        }
        if (tensor_allocate(&input_tensor, 1, 3, H, W) != SUCCESS) {
            free(rgb);
            model_destroy(&model);
            return 1;
        }
        preprocess(&input_tensor, rgb, W, H);
        free(rgb);
        rgb = NULL;
    }

    if (model_load_weights(&model, weights) != SUCCESS) {
        tensor_free(&input_tensor);
        model_destroy(&model);
        fprintf(stderr, "model_load_weights failed\n");
        return 1;
    }

    tensor_t head_output;
    if (tensor_allocate(&head_output, 1, 300, 6, 1) != SUCCESS) {
        tensor_free(&input_tensor);
        model_destroy(&model);
        return 1;
    }

    FILE* out = fopen(out_path, "wb");
    if (!out) {
        fprintf(stderr, "fopen %s failed\n", out_path);
        tensor_free(&input_tensor);
        tensor_free(&head_output);
        model_destroy(&model);
        return 1;
    }

    status_t st = model_forward_ex(&model, &input_tensor, &head_output, out, NULL);
    fclose(out);

    tensor_free(&input_tensor);
    tensor_free(&head_output);
    model_destroy(&model);

    if (st != SUCCESS) {
        fprintf(stderr, "model_forward_ex failed (%d)\n", (int)st);
        return 1;
    }
    printf("Wrote %s\n", out_path);
    return 0;
}
